### PR TITLE
Raise the Netty 5 minimum required Java version to Java 11.

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
Java 11 has been out for some time, and Netty 5 is still some ways out.
There are also many good features in Java 11 that we wish to use, such as VarHandles, var-keyword, and the module system.
There is no reason for Netty 5 to not require Java 11, since Netty 4.x will still be supported for the time being.

Modification:
Remove everything in the pom files related to Java versions older than Java 11.
Remove the animal-sniffer plug-in and rely on the `--release` compiler flag instead.
Remove docker files related to Java versions older than Java 11.
Remove the copied SCTP APIs -- we should test this commit independently on Windows.
Remove the OpenJdkSelfSignedCertGenerator.java file and just always use Bouncy Castle for generating self-signed certificates for testing.

Result:
Java 11 is now the minimum required Java version.

### CODE REVIEW

**Comment by chrisvest** [comment]
**Help wanted:** According to the Git history, all those SCTP files were there to enable that module to compile on Windows. I don't _think_ that's a problem anymore, but I don't have a Windows box to test it on. Can someone who has Java 11 on a Windows machine try compiling this branch (`mvn clean install -DskipTests -T3C` … check the wiki for details) and verify that it still works?

**Comment by chrisvest** [comment]
Hmm… maybe we need to keep a nerfed docker file around to make the centos6-java8 build happy.

**Review by hyperxpro — APPROVED** [review]
I always wanted Netty 5 to use Java 11.

Here are my points to use Java 11:
1. Lots of new APIs will remove a huge amount of boilerplate code.
2. ZGC is available in Linux for super-fast low latency garbage collection.
3. Oracle JDK will support Java 8 till `March 2022` and Java 11 till `September 2023`. However, Amazon Corretto will support Java 8 till `May 2026` and Java 11 till `September 2027`. So we're good for a long time.

The only thing which can hold us back could be: Libraries should be as low as possible but I think it's time to change this. Java 11 is way to go for Netty 5. :D

**Comment by hyperxpro** [comment]
> **Help wanted:** According to the Git history, all those SCTP files were there to enable that module to compile on Windows. I don't _think_ that's a problem anymore, but I don't have a Windows box to test it on. Can someone who has Java 11 on a Windows machine try compiling this branch (`mvn clean install -DskipTests -T3C` … check the wiki for details) and verify that it still works?

Here you go:
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Netty 5.0.0.Final-SNAPSHOT:
[INFO]
[INFO] Netty/Dev-Tools .................................... SUCCESS [  3.477 s]
[INFO] Netty .............................................. SUCCESS [  9.347 s]
[INFO] Netty/Common ....................................... SUCCESS [ 34.246 s]
[INFO] Netty/Buffer ....................................... SUCCESS [ 21.597 s]
[INFO] Netty/Resolver ..................................... SUCCESS [  8.936 s]
[INFO] Netty/Transport .................................... SUCCESS [ 17.874 s]
[INFO] Netty/Codec ........................................ SUCCESS [ 19.902 s]
[INFO] Netty/Codec/DNS .................................... SUCCESS [01:01 min]
[INFO] Netty/Codec/HAProxy ................................ SUCCESS [ 53.009 s]
[INFO] Netty/Handler ...................................... SUCCESS [01:09 min]
[INFO] Netty/Codec/HTTP ................................... SUCCESS [ 36.766 s]
[INFO] Netty/Codec/HTTP2 .................................. SUCCESS [ 59.534 s]
[INFO] Netty/Codec/Memcache ............................... SUCCESS [ 53.429 s]
[INFO] Netty/Codec/MQTT ................................... SUCCESS [ 53.687 s]
[INFO] Netty/Codec/Redis .................................. SUCCESS [ 52.535 s]
[INFO] Netty/Codec/SMTP ................................... SUCCESS [ 52.909 s]
[INFO] Netty/Codec/Socks .................................. SUCCESS [01:00 min]
[INFO] Netty/Codec/Stomp .................................. SUCCESS [ 52.918 s]
[INFO] Netty/Codec/XML .................................... SUCCESS [ 52.019 s]
[INFO] Netty/Handler/Proxy ................................ SUCCESS [ 20.014 s]
[INFO] Netty/Resolver/DNS ................................. SUCCESS [ 22.948 s]
[INFO] Netty/Transport/SCTP ............................... SUCCESS [ 52.285 s]
[INFO] Netty/Example ...................................... SUCCESS [ 36.248 s]
[INFO] Netty/All-in-One ................................... SUCCESS [ 37.345 s]
[INFO] Netty/Transport/Native/Unix/Common ................. SUCCESS [  7.541 s]
[INFO] Netty/Resolver/DNS/MacOS ........................... SUCCESS [  6.516 s]
[INFO] Netty/Tarball ...................................... SUCCESS [  1.687 s]
[INFO] Netty/Transport/Native/Unix/Common/Tests ........... SUCCESS [  4.732 s]
[INFO] Netty/Testsuite .................................... SUCCESS [ 25.646 s]
[INFO] Netty/Transport/Native/Epoll ....................... SUCCESS [ 29.275 s]
[INFO] Netty/Transport/Native/KQueue ...................... SUCCESS [ 25.738 s]
[INFO] Netty/Testsuite/Autobahn ........................... SUCCESS [ 11.876 s]
[INFO] Netty/Testsuite/Http2 .............................. SUCCESS [ 16.452 s]
[INFO] Netty/Testsuite/OSGI ............................... SUCCESS [ 21.354 s]
[INFO] Netty/Testsuite/Shading ............................ SUCCESS [ 12.826 s]
[INFO] Netty/Testsuite/NativeImage ........................ SUCCESS [ 12.510 s]
[INFO] Netty/Transport/BlockHound/Tests ................... SUCCESS [  8.781 s]
[INFO] Netty/Microbench ................................... SUCCESS [01:26 min]
[INFO] Netty/BOM .......................................... SUCCESS [  1.125 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:56 min (Wall Clock)
[INFO] Finished at: 2020-10-07T20:28:49+05:30
[INFO] ------------------------------------------------------------------------
```

**Comment by amizurov** [comment]
**+**, one small note in java 11 we have JNI performance degradation but i'm not tested this for Netty https://bugs.openjdk.java.net/browse/JDK-8210975, maybe it doesn't matter.

**Comment by chrisvest** [comment]
@amizurov That issue looks out of our hands, TBH.

**Comment by normanmaurer** [comment]
+1... I think we should move ahead here

**Comment by hyperxpro** [comment]
@normanmaurer When do we expect first Beta release? Also, can I do some work and contribute in Netty 5? If yes, where can I find suitable TODOs?

**Comment by Scottmitch** [comment]
@netty-bot - test this please

**Comment by Scottmitch** [comment]
master branch is no longer built on jdk8 (tested on https://github.com/netty/netty/pull/10653) and 4.1 will continue to build on jdk8 (tested on https://github.com/netty/netty/pull/10654).

**Comment by amizurov** [comment]
> @amizurov That issue looks out of our hands, TBH.

@chrisvest just FYI https://bugs.openjdk.java.net/browse/JDK-8187809, about possible performance degradation in JNI since Java 10

**Comment by hyperxpro** [comment]
> > @amizurov That issue looks out of our hands, TBH.
> 
> @chrisvest just FYI https://bugs.openjdk.java.net/browse/JDK-8187809, about possible performance degradation in JNI since Java 10

I have tested Netty Epoll 4.1 on Java 8 and Java 11 (Amazon Corretto) and I find Java 11 a little faster. ;/

**Comment by amizurov** [comment]
> > > @amizurov That issue looks out of our hands, TBH.
> > 
> > 
> > @chrisvest just FYI https://bugs.openjdk.java.net/browse/JDK-8187809, about possible performance degradation in JNI since Java 10
> 
> I have tested Netty Epoll 4.1 on Java 8 and Java 11 (Amazon Corretto) and I find Java 11 a little faster. ;/

This not about java 11 slower than Java 8 this only about changes for JNI, and if measured a simple JNI call starting from 10 is slower. It just FYI, maybe this information can helpful because we use a lot of JNI.

**Comment by galderz** [comment]
Would adopting the new select() APIs added via https://bugs.openjdk.java.net/browse/JDK-8199433 make sense for netty? I have doubts those make sense for netty for a couple of reasons: First, netty has an optimization in place to use selected keys that do not generate so much garbage (seems to have started with Trustin in https://github.com/netty/netty/commit/3ce9ab2e72235bf2ae45f57de11803249419cb69). Secondly, the default impl of those methods (see [here](https://github.com/openjdk/jdk11u-dev/blob/master/src/java.base/share/classes/java/nio/channels/Selector.java#L559)) has 2 synchronized blocks in them :(

**Comment by galderz** [comment]
Adopting https://bugs.openjdk.java.net/browse/JDK-6350055 might make more sense, but I wonder whether Netty has ever encountered any similar issue that would require those APIs to be used.

**Comment by normanmaurer** [comment]
@galderz honestly people should just use our "JNI epoll / kqueue transport" or the upcoming "io_uring transport" if they want max performance. So I think it is fine as it is. There I only so much we can do with what the JDK provides as api.

File: docker/docker-compose.centos-6.18.yaml [hunk 1]
```diff
@@ -11,12 +11,15 @@ services:
...
     image: netty:centos-6-1.8
+    command: /bin/echo "skip"
 
   test-leak:
     image: netty:centos-6-1.8
+    command: /bin/echo "skip"
 
   test-boringssl-static:
     image: netty:centos-6-1.8
+    command: /bin/echo "skip"
```
Comment (normanmaurer) [inline docker/docker-compose.centos-6.18.yaml:22]: why do we even need to keep these ?

File: testsuite/pom.xml [hunk 2]
```diff
@@ -94,6 +94,16 @@
...
     </dependency>
+    <dependency>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcpkix-jdk15on</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcprov-jdk15on</artifactId>
+      <scope>compile</scope>
```
Comment (normanmaurer) [inline testsuite/pom.xml:105]: shouldn't these be "runtime" ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: docker/docker-compose.centos-6.18.yaml (hunk 1 above)
Comment (chrisvest) [reply docker/docker-compose.centos-6.18.yaml:22]: Left overs from attempts to make CI happy. I think we can remove that file now.

**Comment by chrisvest** [comment]
@normanmaurer Addressed your comments. Please take a look.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Review by normanmaurer — APPROVED** [review]

**Pull request merged**